### DISCUSSION TITLE
v7.2.0

### DISCUSSION THREAD

**Discussion by Ghost-chu** [PBH-BTN/PeerBanHelper#806] [公告]
与往常一样，我们持续更新 PeerBanHelper 以改进反吸血能力，并持续改进用户体验。  
v7.2.0 是一次次要更新，本次升级涉及配置文件重要变更，请在继续之前备份 `data/config` 文件夹内的文件。

## 旧版本生命周期提醒

由于旧版本中存在的内存泄漏和在线 API 等问题，我们将很快放弃 v7.1.5 版本前的 PeerBanHelper 版本的在线支持。

**如果您仍在使用 v7.1.5 之前的版本，则建议您在 2025 年 1 月 1 日前完成版本升级，更新到最低 v7.1.5 或者更高的 v7.2.0 版本。** 在该日期后，我们将逐步停止旧版本 PeerBanHelper 的在线 API 支持，包括但不限于 GeoIP 库更新和 BTN 网络等服务。
在线 API 停止对旧版本的支持后，旧版本 PeerBanHelper 将无法更新 GeoIP，且也将无法从 BTN 网络获取新的云端规则。
同时由于上面提到的内存泄漏问题，旧版本 PeerBanHelper 可能频繁面临因内存泄漏产生的崩溃问题（在此之前我们通过控制规则数量减缓旧版本的崩溃问题）。

建议您总是保持运行最新版本 PeerBanHelper 以获取重要错误修复和功能更新。

## 反吸血算法改进

改进了对 [BEP-0021 Extension for partial seeds](https://www.bittorrent.org/beps/bep_0021.html) 规范的支持，现在能够对下列情况做出正确的过量下载/进度回退的反吸血判定：

* 种子处于下载状态，部分文件没有完整下载
* 用户/Peer在 “内容” 选项卡中只选择了部分文件下载，而不是整个完整的 torrent (BEP-0021)
* 下载过程中出现错误，导致下载停止，进入了仅上传状态（如 I/O 错误，或者在下载 torrent 过程中磁盘空间不足）

现在 PeerBanHelper 的进度反作弊能够正确处理上面列出的情况。 @paulzzh 

Deluge 用户需要更新[适配器](https://github.com/PBH-BTN/PBH-Adapter-Deluge/releases)

## 通知渠道管理

现在可在 WebUI 上管理通知推送渠道。当出现警告或者错误的高等级信息时（如：上传量达到每日告警阈值；PBH 内存不足等），您将会收到通知推送。 @Gaojianli 

![image](https://github.com/user-attachments/assets/88b1760a-4dd3-492d-8bac-8ad9e051902c)

![image](https://github.com/user-attachments/assets/183c8b05-6c6c-40d8-893c-ab99ca3ffc4f)

## 可显示/隐藏的图表项

现在图表页面可选择显示或者隐藏特定的图表。@Gaojianli

![image](https://github.com/user-attachments/assets/f90ab8ba-d11c-45bd-bee0-b29d267325cf)

## PeerBanHelper Labs

欢迎来到 PeerBanHelper 实验室！我们有时会对 PeerBanHelper 的一些功能做出实验性的更改。这些实验正在进行中，有的则可能被放弃。我们希望其中的多数会获得您的反馈意见，得以改进，并最终成为 PeerBanHelper 的一部分。PeerBanHelper 实验室的作用就在于此。

每个实验推出时都会随机的在一部分用户的 PeerBanHelper 上启用，以便获得新功能的反馈。但如果新功能的对您造成了困扰/或者想要提前尝鲜测试，您也可以随时在 “实验室” 菜单中手动开启关闭实验项目。@Ghost-chu @Gaojianli 

![PeerBanHelper Labs](https://github.com/user-attachments/assets/371f772b-2b7d-481c-aa44-7983503d889f)

## install4j 自动更新支持

被大家心心念念的 Windows/Linux 平台的 install4j 自动更新的支持现已到来。对使用安装程序安装的 PeerBanHelper，在本次更新后将会添加自动更新功能。每次启动 PeerBanHelper 时，更新程序会在后台检查并下载更新，并在下一次启动时安装新的更新。

您也可以通过开始菜单中新增的 “PeerBanHelper 检查更新” 程序来手动检查更新：

![image](https://github.com/user-attachments/assets/378ea19a-f020-43b9-93cd-9c11f9fd08f0)

![de9aa83ed64e53850d3abcc10b49eb35](https://github.com/user-attachments/assets/30006a7c-63cd-4856-a7cd-210e1ca6e77c)

对于非 install4j 安装程序安装的 PeerBanHelper：

* Docker：可使用 WatchTower 自动更新
* 群晖套件：继续由群晖的套件中心提供更新
* .deb 包：由用户自行管理更新

@Ghost-chu 

## 更新日志

现在可以直接在前端查看新版本的更新日志 @Gaojianli 
![image](https://github.com/user-attachments/assets/eaf9d699-e365-4357-8fe1-8053cbfea37e)
![image](https://github.com/user-attachments/assets/3a860e79-49d7-4c55-9070-d9232f70d246)

## （实验性功能）IPFS/IPNS 规则订阅支持

> [!NOTE]
> 这是一个实验性功能，不建议普通用户使用

作为 #622 的一部分，本次 v7.2 版本中新增了 IPFS/IPNS 去中心化网络支持。[IPFS（星际文件系统）](https://ipfs.tech/) 是一个类似 BitTorrent 的完全去中心化点到点超媒体传输协议，而 [IPNS](https://docs.ipfs.tech/concepts/ipns/) 则充当 IPFS 网络中的 DNS。

要使用 IPFS/IPNS 规则订阅，则需要部署 Kubo IPFS 客户端，在实验室中启用 IPFS 实验并在 config.yml 手动启用和配置。配置完成后，您就可以使用 IPFS 和 IPNS 协议地址来添加订阅。

您可以在配置文件中修改 Kubo RPC 地址为您自行安装的 Kubo IPFS 客户端，或者使用新的 docker-compose.yml 部署，将会自动配置 IPFS 客户端。

* 对于 IPFS 来说，你需要使用 `ipfs://CID` 格式的 URL，但缺点是每个文件被提交到 IPFS 后都无法再次被修改
* 您也可以使用 IPNS，IPNS 是可修改的，并可以指向任意 CID，格式为 `ipns://ID`，你可以在 IPFS Desktop 中将不同的文件发布到同一个 IPNS 上来完成更新操作

同时，PeerBanHelper 也会每隔 1 个小时（可修改）将自身的 Banlist 发布到 IPFS 和 IPNS 上，IPNS 地址可在 IPFS 客户端上看到。如果您的 IPFS 客户端连接了多个 PeerBanHelper 客户端，则您可以通过后缀（安装ID）来识别不同的 PeerBanHelper 客户端。

![Kubo Client IPNS](https://github.com/user-attachments/assets/9067943f-9451-4df8-9065-a9b27378e8f5)

使用 IPFS 网络可以做到人人都能创建合订阅规则，且无需购买服务器。使用的人越多，则帮助你一起分发规则的人也越多。 @Ghost-chu 

## DEB 安装包改进

感谢 @Anuskuss。本次更新后，使用 DEB 安装包安装 PeerBanHepler 的用户将会将数据文件和 PBH 主程序安装到更合适的位置。本次更新也同时改进了日志记录功能。#732

## PBH Plus 许可证更改

在过去只有在爱发电上捐赠我们的用户可以使用 PBH Plus 的捐赠功能，但我们认为这可能有悖于非营利软件的定义。因此在 v7.2 版本中，我们添加了一个 “我不想捐赠” 的按钮，允许直接激活 PBH Plus 许可证而无需捐赠。

通过这种方式激活的 PBH Plus 许可证在功能上与捐赠用户的许可证完全相同，但有效期只有 15 天，且不会显示捐赠感谢信息。当许可证到期后，您可以重复相同步骤，再次点击 “我不想捐赠” 创建新的循环许可证。

![image](https://github.com/user-attachments/assets/491b9116-6c23-42b1-b410-6f1e9dae0538)

尽管如此，PBH-BTN 仍在维护多个需要持续投入的基础设施，如 BTN 和公共 BitTorrent Tracker。  
如果 PeerBanHelper 确实有帮到您，我们仍然希望您能够考虑捐赠此项目。

## 错误修复

* 修复了 BTN、检查间隔 等配置项在 WebUI 调整设置后重载配置不生效的问题 @Ghost-chu
* 修复了新安装首次启动时清理任务报告 SQL 错误的问题 @Ghost-chu 
* 修复端口占用时，弹出的提示没有显示被占用的端口号的问题 @Ghost-chu  

## 性能优化

* 改进了加载 BTN 规则时产生的内存尖峰，避免极端情况下可能导致 OOM 的问题 @Ghost-chu 
* 改进了例外列表 IP 判断的性能和内存占用 @Ghost-chu 

## 用户体验优化

* 安装向导结束时，添加 “启动 PeerBanHelper” 复选框 @Ghost-chu 
* 连接到 qBittorrent 下载器时，在首次登陆时自动关闭需要关闭的 “允许同一 IP 地址的多重连接” 开关 @Ghost-chu 

## 文档修正

* 修正了多个文档中的拼写错误 @CreeperAWA 

### Sparkle BTN 改进

自上一个版本发布以来，我们对 Sparkle BTN 做了大量的改进，以提升其分析精确度和分析性能。

* BTN 对提交的数据进行分析时，现在按照时间块进行切分来统计数据，以避免短时间内同一 Peer 触发多个 BTN 客户端，导致连续上报被立刻标记为不可信的问题
* 将统计时间区间从 45 天缩短到 14 天。由于动态 IP 地址的特性，一个 IP 很少能够被同一个用户持续使用超过 7 天。对于普通用户来说，这改善了误封禁的问题；对于数据中心 IP 来说，因为其 IP 一直没有变化，因此这个调整不会对其产生影响；对于动态吸血 IP 来说，它们遵循与普通用户一样的 IP 轮换规则，因此 IP 被污染抛弃后，BTN 现在可以加快过期速度，释放原先的 IP 地址
* 与 Sparkle Tracker 联动

### Sparkle Tracker 改进

* 大幅改进了 Sparkle Tracker 的性能，解决了因繁重的 Tracker 任务导致的多次 Sparkle 程序宕机、崩溃和数据库服务器宕机的问题
* 现在支持与 Sparkle BTN 服务交叉联动，BTN 现在能够分析 Sparkle Tracker 的数据以生成反吸血 IP 地址规则
* 后期将进一步深度集成两者，以便聚合分析数据

> [!NOTE]
> 小提示：如果您是种子发布者，您可以将 Sparkle Tracker (`https://sparkle.ghostchu-services.top/announce`) 加入到您的种子中，这可以诱导恶意 Peer 连接到 Sparkle Tracker 以便 Sparkle BTN 进行特征分析。
> 同时，Sparkle Tracker 也是一个公共 BitTorrent Tracker，可以提供正常的 Tracker 服务。

## 回退支持

v7.2.0 是一次次要更新，因此支持版本回退。目前支持的回退路径为回退到最低 v7.1.5。  
您可通过还原 v7.1.5 的 `data/config` 中的文件备份后，重新安装 v7.1.5 版本以降级到 v7.1.5。

## Docker

DockerHub: `ghostchu/peerbanhelper:v7.2.0`  
阿里云国内镜像加速: `registry.cn-hangzhou.aliyuncs.com/ghostchu/peerbanhelper:v7.2.0`

<hr /><em>This discussion was created from the release <a href='https://github.com/PBH-BTN/PeerBanHelper/releases/tag/v7.2.0'>v7.2.0</a>.</em>

**Comment by Mishasama** [comment]
不是很理解为什么对额外使用了更多`非必要且有额外使用成本的`服务器资源的用户要求支付其使用成本属于盈利行为。
难道之前发布 Plus 时对其收取费用的说明是虚构的？还是说你们确实提高了价格，导致出现了差价盈利？

你可以提供免费解锁的“促销”活动，但说可能有悖于非盈利定义我觉得有点不妥。<s>请公开账单以证明你们没有盈利。</s>

**Reply by Ghost-chu** [reply]
> 不是很理解为什么对额外使用了更多`非必要且有额外使用成本的`服务器资源的用户要求支付其使用成本属于盈利行为。 难道之前发布 Plus 时对其收取费用的说明是虚构的？还是说你们确实提高了价格，导致出现了差价盈利？
> 
> 你可以提供免费解锁的“促销”活动，但说可能有悖于非盈利定义我觉得有点不妥。请公开账单以证明你们没有盈利。

PBH Plus 的设计初衷有如下几点：

1. 支付基础设施费用。这毫无疑问，但实际上并不多，每个月实际开销大概在 60-150 CNY 左右。这看起来并不多但需要特别注意的是，相比于一次性捐赠，这些费用是持续支出的。除非有源源不断的新用户，否则实际上是用未来的钱支付现在的账单。因此成本控制较为重要。
2. 用于支持开发人员。除了一部分 “保留” 的资金（这部分资金保留在 PBH-BTN 公共资金池内用作持续开支，我们一般会保留约 1 年的预算），剩余的将作为奖励分发给 PBH-BTN 组织的正式成员，目前这部分成员共计 5 人。具体金额根据当月补充公共资金池后剩余的部分多少而有所不同。如果持续开支出现变动（如因新购服务器或资源包而增加），则下个月的发电金额将补充保留部分的差额，直至保留部分恢复正常水平。开发人员获取这部分奖励金额后可自行决定如何使用它们，只要按时完成版本计划和功能需求即可。
3. 用于非个人的其他用途，但需要得到批准。例如：升级群人数、捐赠其它项目等。只要有利于组织、BitTorrent 社区等均是可以的。

对一些问题的回答：

> 不是很理解为什么对额外使用了更多非必要且有额外使用成本的服务器资源的用户要求支付其使用成本属于盈利行为。

我不是很理解这个问题，但实际情况是我们的主要目标是淘汰 v6 版本，因为大量的钉子户迟迟没有升级 v7 版本。在规则更新后，v6 版本的内存消耗频繁达到上限导致崩溃，引发大量支持请求，因此主要目的是淘汰 v6 版本，而不是减少服务器运行成本。  
我们目前使用 4 台机器运行 BTN 网络：

* 一台实体服务器：在项目初期用于 BTN 运行，现在用于流复制备份数据。在代码重构后我们计划将计算任务重新移动回这台设备上
* 一台 Oracle Always Free 服务器：目前运行 BTN 网络的主力，同时运行了 [Trunker](https://github.com/PBH-BTN/trunker) 以提供公共 Tracker 服务。它唯一的问题是 CPU 和磁盘 I/O 性能较差，在高负载时频繁出现卡死的问题
* 一台腾讯云的国内服务器：主要用于中继国内用户的流量，解决部分用户与 CloudFlare 通信不顺畅的问题。在 BTN 挂掉的时候也承担显示错误信息的功能
* 一台 Synology NAS：在我家里，主要负责上面所有的设备的数据备份

因此实际上，运行成本并不高。

> 难道之前发布 Plus 时对其收取费用的说明是虚构的？还是说你们确实提高了价格，导致出现了差价盈利？

Plus 发布时对捐赠的说明可在旧版 WebUI 仓库中找到：[跳转链接](https://github.com/PBH-BTN/pbh-fe/blob/fe61bb64ea4c0d19e0970b48a155b769e4606e72/src/locale/zh-CN/plus.ts#L12)，此处没有对用途做出保证和说明。

捐赠页面至今仍在使用且没有作出修改：[爱发电页面](https://afdian.com/item/bf4dae245bb011efa1ca52540025c377)

实际上，除本次更新外，PBH-BTN 从未发表过有关非盈利、非营利的文字。您可以自行查找相关资料。

   
PBH Plus 的设立初衷是对捐赠用户解锁特定功能以鼓励用户捐赠，但这实际上是一种功能性的付费墙，它违反了 PBH 正在使用的开源许可证。  
因此在 v7.2 中将其转为免费的以避免与开源许可证冲突，同时也是一种对没有捐赠方式的用户提供获取与捐赠用户相同体验的手段。 
你需要知道的是，世界上的一切都是持续付费的，但许可证是一次性，没有到期时间。

至于 “非必要且有额外使用成本”，其实并不会造成额外成本。v6 的钉子户用户长期驻留在存在问题的版本，服务端不得不一直缓解这些用户提交的重复数据，除此以外，推动彻底放弃 v6 也是因为 v6 的卡死崩溃和内存溢出问题的存在。v7 解决了上述问题，v6 不再有任何优点值得继续使用，BTN 也可以放弃对 v6 数据结构的特殊转换，改善可维护性。这是一个固定的过程，就像 v6 推出后丢弃了 v5 的支持一样。

> 你可以提供免费解锁的“促销”活动，但说可能有悖于非盈利定义我觉得有点不妥。

15 天的倒计时设计目的：毫无疑问的，主要设计目的是鼓励用户捐赠，因为这直接影响项目的可持续性。如果将其设计为点击一次再也不会弹出，这显然会让许多人忘记它。

你有很多办法绕过它：

1. 你可以从别的地方复制一串捐赠密钥然后直接使用。PBH 不会联网校验，因此这完全可行，也是受到我们认可的方式
2. 你可以破解 PeerBanHelper，直接修改代码，去除 PBH Plus 检查的部分。相关代码是开源的一部分，这同样受到我们的认可
3. 【推荐】直接向 PBH-BTN 组织的项目提交高价值补丁，成为 PBH-BTN 社区的成员。你不但可以随便签发许可证，还可以查询支付宝的账单情况。当然作为正式成员的一员，你自然也会参与捐赠奖励计划之中

但无论如何，是否捐赠都不会影响你使用全部功能。

---

一切捐赠的唯一目的：保证项目的活性。让现有的社区成员不离开且坚持下来。这是一个漫长的过程且需要一直持续下去。  
热情会消散，意志也会磨灭。用爱发电不会长久，这也是捐赠奖励的意义所在——作为纯爱好以外的现实世界中的物质上的支持。  

PeerBanHelper 与其它程序不同的是，作为一个对抗工具，它需要不断地更新。道高一尺魔高一丈，军备竞赛从不停止。PBH 之所以效果最佳的原因是，背后是大量的人力堆砌。这与每一个深度参与开发的人密不可分。这也是我做出这个决定的原因，PBH-BTN 社区的成员应当得到他们付出的回报。  
捐赠与否对我个人并不重要，有没有都行。一分没有这么多年该走下来的也走下来了。但我不应该也以同样的要求来要求社区的其他成员。

如果层主对 PBH-BTN 的捐赠相关问题有更多建议，你可以通过加入 PBH-BTN 社区正式成员的方式参与其中。一旦您成为正式成员组的一员，自然也有权管理、使用和监督捐赠资金。以下是一些成为正式成员的方式：

* 提交高价值功能补丁，前后端均可
* 作为 DBA 加入，帮助维护数据库的正常运行，并提出查询改进意见
* 成为 Review 组的成员，在版本发布前审查和测试 PR 中的代码，测试并找出其中的缺陷
* 作为采购加入 PBH-BTN，负责计算和网络资源的采购，如管理 PBH-BTN 的基础设施和云资源的账单等，关注各个云厂商的折扣活动等，进行成本控制
* 作为运维加入 PBH-BTN，负责计算和网络资源的维护，如即时续签证书，在服务挂掉的时候及时恢复
* 作为规则管理源加入 PBH-BTN，负责 BTN 规则的人工部分的维护，通过连接生产库查询的方式，更新 Github 上的规则等

再次感谢层主对 PBH-BTN 组织项目的关注。

**Reply by Mishasama** [reply]
感谢如此耗费时间精力进行长文回复，这有点超出我的预期。😂
- 首先，支付给项目的工作人员的“费用”属于成本、开销，不属于盈利。而且该费用也应该没有超出合理范围。（除非你确定每个人每个月**稳定**能获得社会平均工资以上的收入，否则它不属于“工资”，顶多算是“报销”。而“报销”**依法是属于免税不计入营收所得**的）
- 其次，Plus的功能依赖的服务器和维护人员存在持续开销的成本。要求用户支付（摊分）使用成本同样不属于盈利行为。而且正如你说的，捐赠可能是一次性的行为，而开销是长期持续存在的，短时间可能不多，但一旦项目持续运作，那就很有可能出现赤字。所以用户支付的甚至都不能称作捐赠，而是**使用该功能的所必须支付的个人成本**。
  - 即便他们为了这个功能而自行架设对应的服务器也一样有这样的成本，难道他们自己给自己“捐赠”了？
  - 所以实际上这并不是付费墙，这是给用户一个方便租用该功能所需的服务器的其中一种方式而已。**为了解决许可证冲突的疑虑，你只需要给出一个选项给用户修改要连接到的服务器地址就可以了。** 项目提供的公益服务器属于摊分使用成本且免去自行架设和维护的模式而已。
- 最后，对于要推进版本迭代，根据你的描述，我认为只要在v7中更换服务器的地址即可。让v6中指向的服务器无效，他们就不得不更新或者寻求其它办法了。这样服务器也不会被“DDoS”了。
----
> 如果层主对 PBH-BTN 的捐赠相关问题有更多建议，你可以通过加入 PBH-BTN 社区正式成员的方式参与其中。

我不是专业程序员无法在代码相关的工作上提供帮助。
但如果项目遇到需要公关、杠法律/道德问题、运营策略等，或者UX，我都可以提供支持。
团队有精神/心理建设方面的需求也行……<s>但生理需求请恕我无能为力。</s>

### 其实我只是看到你突然这么一说，担心项目是否被人杠上了而做出妥协行为而已。**我个人不赞成公益项目向杠精/利益集团低头。**
